src/java/main/org/apache/zookeeper/server/WatchManager.java

```diff
@@ -97,24 +157,47 @@ synchronized void removeWatcher(Watcher watcher) {
     Set<Watcher> triggerWatch(String path, EventType type, Set<Watcher> supress) {
         WatchedEvent e = new WatchedEvent(type,
                 KeeperState.SyncConnected, path);
-        Set<Watcher> watchers;
+        Set<Watcher> watchers = new HashSet<>();
         synchronized (this) {
-            watchers = watchTable.remove(path);
-            if (watchers == null || watchers.isEmpty()) {
-                if (LOG.isTraceEnabled()) {
-                    ZooTrace.logTraceMessage(LOG,
-                            ZooTrace.EVENT_DELIVERY_TRACE_MASK,
-                            "No watchers for " + path);
+            PathParentIterator pathParentIterator = getPathParentIterator(path);
+            for (String localPath : pathParentIterator.asIterable()) {
+                Map<Watcher, Type> thisWatchers = watchTable.get(localPath);
+                if (thisWatchers == null || thisWatchers.isEmpty()) {
+                    continue;
                 }
-                return null;
-            }
-            for (Watcher w : watchers) {
-                Set<String> paths = watch2Paths.get(w);
-                if (paths != null) {
-                    paths.remove(path);
+                Iterator<Entry<Watcher, Type>> iterator = thisWatchers.entrySet().iterator();
+                while (iterator.hasNext()) {
+                    Entry<Watcher, Type> entry = iterator.next();
+                    Type entryType = entry.getValue();
+                    Watcher watcher = entry.getKey();
+                    if (entryType.isRecursive()) {
+                        if ( type != EventType.NodeChildrenChanged ) {
```

Comment:
Maybe I am dumb to ask this, but why watchEvent type could not be NodeChildrenChanged here?